pkg/retrieval/retrieval.go

```diff
@@ -90,32 +98,85 @@ func (s *Service) RetrieveChunk(ctx context.Context, addr swarm.Address) (swarm.
 		span, logger, ctx := s.tracer.StartSpanFromContext(ctx, "retrieve-chunk", s.logger, opentracing.Tag{Key: "address", Value: addr.String()})
 		defer span.Finish()
 
-		var skipPeers []swarm.Address
+		sps := newSkipPeersService()
 
-	LOOP:
-		for i := 0; i < maxPeers; i++ {
-			select {
-			case <-ctx.Done():
-				break LOOP
-			default:
+		ticker := time.NewTicker(retrieveRetryIntervalDuration)
+
+		var (
+			peerAttempt int
+			resultC     = make(chan retrieveChunkResult)
+			wgChan      sync.WaitGroup
+		)
+
+		wgChan.Add(maxPeers)
+
+		defer func() {
+			if peerAttempt != maxPeers {
+				wgChan.Add(peerAttempt - maxPeers)
+			}
+			wgChan.Wait()
+
+			close(resultC)
+		}()
+
+		for {
+			if peerAttempt < maxPeers {
+				peerAttempt++
+
+				go func() {
+					defer wgChan.Done()
+
+					chunk, peer, err := s.retrieveChunk(ctx, addr, sps)
+					if err != nil {
+						if peer.IsZero() {
+							select {
+							case resultC <- retrieveChunkResult{err: err}:
+							default:
+							}
+							return
+						}
+
+						logger.Debugf("retrieval: failed to get chunk %s from peer %s: %v", addr, peer, err)
+
+						select {
+						case resultC <- retrieveChunkResult{err: storage.ErrNotFound}:
+						default:
+						}
+						return
+					}
+
+					select {
+					case resultC <- retrieveChunkResult{chunk: chunk}:
+					default:
+					}
+				}()
+			} else {
+				ticker.Stop()
 			}
 
-			var peer swarm.Address
+			select {
+			case <-ticker.C:
+				// break
+			case result := <-resultC:
+				if result.err != nil {
```

Comment:
If the error is received here we do fail, however as it can be seen in the goroutine that is executing actual request (whose flow was not changed), only when `peer.IsZero()` is the error returned (and whole thing stopped). Otherwise we just log that we failed, and return.

One addition I made there is if that happens error is returned, more specifically `storage.ErrNotFound` error, which is handled here to break `select`, which should cause another loop, which will potentially cause another request toe some other peer.

Reply:
So currently (on master), if the function that does the retrieval returns an error, then the next peer will be tried, but if I understand correctly, in this implementation, assuming that multiple goroutines are trying to retrieve the chunk and the first one returns an error - the whole thing will then fail, although the chunk can still arrive from another peer. Is this correct or am I reading this wrong?

Reply:
Looking at current code on `master`, if retrieval returns an error (specifically error which includes `peer.IsZero()`) then the whole process stops, and no further peers will be contacted. The changes I made are suppose to be the same.

What I may have not explored in more detailed is when exactly will `peer` which is "zero"  will be returned.
It might be the case this is just when we no longer have more peers to which node is connected to, in which case this check should be removed altogether.